Submodule src/boringssl contains modified content
diff --git a/src/boringssl/CMakeLists.txt b/src/boringssl/CMakeLists.txt
index dad27f815..11e517b52 100644
--- a/src/boringssl/CMakeLists.txt
+++ b/src/boringssl/CMakeLists.txt
@@ -499,12 +499,26 @@ else()
   message(FATAL_ERROR "Unknown processor:" ${CMAKE_SYSTEM_PROCESSOR})
 endif()
 
-if(ANDROID AND NOT ANDROID_NDK_REVISION AND ARCH STREQUAL "arm")
-  # The third-party Android-NDK CMake files somehow fail to set the -march flag
-  # for assembly files. Without this flag, the compiler believes that it's
-  # building for ARMv5.
-  set(CMAKE_ASM_FLAGS "-march=${CMAKE_SYSTEM_PROCESSOR} ${CMAKE_ASM_FLAGS}")
-endif()
+# set ARCH and MARCH
+set(ARCH ${TARGET_ABI})
+set(MARCH ${ARCH})
+if(TARGET_ABI STREQUAL "arm")
+    set(MARCH "armv7-a")
+elseif(TARGET_ABI STREQUAL "aarch64")
+    set(MARCH "armv8-a")
+elseif(TARGET_ABI STREQUAL "x86")
+    set(MARCH "i686")
+elseif(TARGET_ABI STREQUAL "x86_64")
+    set(MARCH "x86_64")
+endif()
+
+# set -march flag
+ if(ANDROID AND NOT ANDROID_NDK_REVISION AND ARCH STREQUAL "arm")
+   # The third-party Android-NDK CMake files somehow fail to set the -march flag
+   # for assembly files. Without this flag, the compiler believes that it's
+   # building for ARMv5.
+  set(CMAKE_ASM_FLAGS "-march=${MARCH} ${CMAKE_ASM_FLAGS}")
+ endif()
 
 if(USE_CUSTOM_LIBCXX)
   if(NOT CLANG)
