Submodule src/libpng contains modified content
diff --git a/src/libpng/CMakeLists.txt b/src/libpng/CMakeLists.txt
index 6451fcf1b..bccd2ef74 100644
--- a/src/libpng/CMakeLists.txt
+++ b/src/libpng/CMakeLists.txt
@@ -36,18 +36,18 @@ include(GNUInstallDirs)
 
 # Allow users to specify location of Zlib.
 # Useful if zlib is being built alongside this as a sub-project.
-option(PNG_BUILD_ZLIB "Custom zlib Location, else find_package is used" OFF)
+option(PNG_BUILD_ZLIB "Custom zlib Location, else find_package is used" ON)
 
 if(NOT PNG_BUILD_ZLIB)
   find_package(ZLIB REQUIRED)
   include_directories(${ZLIB_INCLUDE_DIR})
 endif()
 
-if(UNIX AND NOT APPLE AND NOT BEOS AND NOT HAIKU)
+if(UNIX AND NOT APPLE AND NOT BEOS AND NOT HAIKU AND NOT ANDROID)
   find_library(M_LIBRARY m)
 else()
   # libm is not needed and/or not available
-  set(M_LIBRARY "")
+  set(M_LIBRARY "m")
 endif()
 
 # COMMAND LINE OPTIONS
